---
title: "Host Transcriptome"
author: "John Sterrett"
date: "2023-10-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load("tidyverse",
               "ggplot2")
# set ggplot theme
theme_set(theme_bw())
```

# Load data
```{r}
gn <- data.table::fread("../seq.f0.0.r0.0.host/counts.txt", data.table=F)
colnames(gn) <- colnames(gn) %>% gsub(pattern="seq.f0.0.r0.0.host/",
                                      replacement="")
colnames(gn) <- colnames(gn) %>% gsub(pattern=".bam",
                                      replacement="")
rownames(gn) <- gn$Geneid
gn$Geneid <- NULL

gns.only <- gn[!colnames(gn) %in% c("Chr", "Start", "End", "Strand", "Length")] %>%
    as.matrix()



metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)
```

# Convert to transcripts per million (tpm)
```{r}
# https://support.bioconductor.org/p/91218/
# convert to TPM
tpm <- function(counts,len) {
  x <- counts/len
  return(t(t(x)*1e6/colSums(x)))
}

gns.only.tpm <- tpm(gns.only, gn$Length)
gns.only.tpm.pseudo <- tpm(gns.only+1, gn$Length)

```

# Ordination {.tabset}
```{r}
tpm.aitch <- robCompositions::aDist(t(gns.only.tpm.pseudo))
tpm.aitch.pcoa <- ape::pcoa(tpm.aitch)


tpm.aitch.pcoa.dat <- merge(tpm.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")

```

## Plot by HIV Status {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

Does host transcriptome correlate with metatranscriptome and/or function?

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4)
```

## Plot by Cohort {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=Cohort, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=Cohort, label=Row.names)) +
    geom_point(size=4)
```

## Plot by storage method {.tabset}
```{r}
# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

tpm.aitch.pcoa.dat$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
tpm.aitch.pcoa.dat[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
tpm.aitch.pcoa.dat[is.na(tpm.aitch.pcoa.dat$storage_method), "storage_method"] <- "RNALater"

```

### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4)
```
