---
title: "Host Transcriptome"
author: "John Sterrett"
date: "2023-10-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load("tidyverse",
               "ggplot2",
               "DESeq2",
               "EnhancedVolcano",
               "pathfindR",
               "fgsea",
               "IHW")

# set ggplot theme
theme_set(theme_bw())
dir.create("output", showWarnings = FALSE)
```

# Load data
```{r}
gn <- data.table::fread("../seq.f0.0.r0.0.host/counts.txt", data.table=F)
colnames(gn) <- colnames(gn) %>% gsub(pattern="seq.f0.0.r0.0.host/",
                                      replacement="")
colnames(gn) <- colnames(gn) %>% gsub(pattern=".bam",
                                      replacement="")
rownames(gn) <- gn$Geneid
gn$Geneid <- NULL

gns.only <- gn[!colnames(gn) %in% c("Chr", "Start", "End", "Strand", "Length")] %>%
    as.matrix()



metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)
```

# Convert to transcripts per million (tpm)
```{r}
# https://support.bioconductor.org/p/91218/
# convert to TPM
tpm <- function(counts,len) {
  x <- counts/len
  return(t(t(x)*1e6/colSums(x)))
}

gns.only.tpm <- tpm(gns.only, gn$Length)
gns.only.tpm.pseudo <- tpm(gns.only+1, gn$Length)

write.csv(gns.only.tpm, "../seq.f0.0.r0.0.host/host_gene_counts_tpm.csv")

```

# Ordination {.tabset}
```{r}
tpm.aitch <- robCompositions::aDist(t(gns.only.tpm.pseudo))
tpm.aitch.pcoa <- ape::pcoa(tpm.aitch)


tpm.aitch.pcoa.dat <- merge(tpm.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")

```

## Plot by HIV Status {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

Does host transcriptome correlate with metatranscriptome and/or function?

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4)
```

## Plot by Cohort {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4)
```

## Plot by storage method {.tabset}
```{r}
# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

tpm.aitch.pcoa.dat$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
tpm.aitch.pcoa.dat[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
tpm.aitch.pcoa.dat[is.na(tpm.aitch.pcoa.dat$storage_method), "storage_method"] <- "RNALater"

```

### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4)
```

## PERMANOVA on Aitchison distance

### Prep data

We'll need to only grab timepoint 1 data to avoid longitudinal sampling from the same individual
```{r}
get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

# subset the distance matrix to only time 1
index.t1 <- which(get_timepoint(names(tpm.aitch))==1)
names.t1 <- names(tpm.aitch)[index.t1]
t1.tpm.aitch <- usedist::dist_subset(tpm.aitch, names.t1)

# Subset the metadata
t1.tpm.aitch.pcoa.dat <- tpm.aitch.pcoa.dat[index.t1,]

# Make sure it worked
if(sum(t1.tpm.aitch.pcoa.dat$Row.names==names(t1.tpm.aitch)) != nrow(t1.tpm.aitch.pcoa.dat)){
    stop("order of dist matrix and df are not the same")
}
```

### Run PERMANOVA on timepoint 1 only as a function of storage method
```{r}
vegan::adonis2(t1.tpm.aitch ~ t1.tpm.aitch.pcoa.dat$storage_method, 
               permutations=10^5)

```

# Differential Expression {.tabset}
## Prep data
```{r DESeq_prep}
metadata <- merge(metadata,
                  tpm.aitch.pcoa.dat[c("Row.names", "storage_method")], 
                  by.x="Sample_ID", by.y="Row.names")


metadata$timepoint <- get_timepoint(metadata$Sample_ID)

subset_metadata <- metadata %>% filter(storage_method=="RNALater",
                                       timepoint==1)
gns.only.subset <- gns.only[,subset_metadata$Sample_ID]
```

## HIV Status
### Run DESeq2
```{r DESeq_run}
ds.dat <- DESeqDataSetFromMatrix(countData=gns.only.subset, 
                                 colData=subset_metadata,
                                 design=~HIV_Status)

# filter low count genes
keep <- rowSums(counts(ds.dat)) >= 10
ds.dat <- ds.dat[keep,]

dds <- DESeq(ds.dat)
res <- results(dds, contrast=c("HIV_Status", "Positive", "Negative"), 
               pAdjustMethod="fdr")
res.ape <- lfcShrink(dds=dds, coef=2, type="apeglm")
# https://support.bioconductor.org/p/9141271/
#res.ape <- lfcShrink(dds=dds, coef=2, type="apeglm", 
#                     lfcThreshold=log(1.5, base=2),
#                     svalue=FALSE)

#resIHW <- results(dds, filterFun=ihw)
#summary(resIHW)
#resIHW[resIHW$padj < 0.1,]
#metadata(resIHW)$ihwResult


res <- res[order(res$padj),] 
res.ape <- res.ape[order(res.ape$padj),] 

res %>% as.data.frame() %>% head(n=10) %>% 
    knitr::kable(caption="Top DEGs")
res.ape %>% as.data.frame() %>% head(n=10) %>% 
    knitr::kable(caption="Top DEGs with LFC shrinkage applied")

res %>% write.csv("output/NOTadjusted_DESeq2results_HIV.csv")
res.ape %>% write.csv("output/APE_adjusted_DESeq2results_HIV.csv")
```

### MA Plot {.tabset}

#### Apeglm shrunken LFC
```{r ma_shrunk, fig.height=6}
plotMA(res.ape, ylim=c(-2,2))

```

#### Raw LFC
```{r ma_raw, fig.height=6}
plotMA(res, ylim=c(-2,2))

```


### Volcano plot {.tabset}
Plots show unadjusted p values. The bonferroni-adjusted p value threshold is between 5 and 6 on the y axis, and if we do a Benjamini-Hochberg correction, again only genes with p < 10^-5 are significant. 

To the left indicates higher expression in HIV(-) individuals, and to the right indicates higher expression in HIV(+) individuals.

#### Apeglm shrunken LFC
```{r DESeq_volcano_shrunk, fig.height=8}
EnhancedVolcano(res.ape,
                lab = rownames(res.ape),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```

#### Raw LFC
```{r DESeq_volcano_raw, fig.height=8}
EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```


### pathfindR to map gene ids to pathways
PathfindR errors because it can't really do anything with the 2 differentially expressed genes.
```{r pathfindR, eval=FALSE}
# Grab the relevant columns from the DESeq2 results
pfr.in <- res[,c("log2FoldChange", "padj")] %>% as.data.frame()
pfr.in$Geneid <- rownames(pfr.in)

# order columns to be "Geneid", "log2FoldChange", "padj"
pfr.in <- pfr.in[,c(3,1,2)]

# remove NA p values
pfr.in <- pfr.in %>% filter(is.na(padj)==F)

pfr.res <- run_pathfindR(pfr.in, output_dir="pathfindR",
                         visualize_enriched_terms=F, plot_enrichment_chart=F)

```

### GSEA {.tabset}
```{r}
res2 <- res.ape %>% as.data.frame() %>%
  arrange(pvalue, log2FoldChange) %>%
  select(log2FoldChange) %>% 
  na.omit()

res2$SYMBOL <- rownames(res2)
# create a named vector of the gene stats
ranks <- tibble::deframe(res2[c("SYMBOL", "log2FoldChange")])
# read in Hallmark pathways, https://www.gsea-msigdb.org/gsea/msigdb
pathways.hallmark <- gmtPathways("../h.all.v2023.2.Hs.symbols.gmt")
pathways.c2 <- gmtPathways("../c2.cp.v2023.2.Hs.symbols.gmt")
pathways.c7.immune <- gmtPathways("../c7.immunesigdb.v2023.2.Hs.symbols.gmt")
pathways.celltype <- gmtPathways("../c8.all.v2023.2.Hs.symbols.gmt")

```
#### Hallmark
```{r gsea, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)


fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  write.csv("output/GSEA_Hallmark_HIV.csv")
```

#### Curated Gene Sets
```{r gsea.c2, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.c2, stats=ranks)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)

fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  write.csv("output/GSEA_Curated_HIV.csv")

```

#### Immune
```{r gsea.immune, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.c7.immune, stats=ranks)


fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)

fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>%
  write.csv("output/GSEA_Immune_HIV.csv")
```

#### Cell types
```{r gsea.celltype, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.celltype, stats=ranks)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)

```


## MSM Status
### Run DESeq2
```{r DESeq_run.msm}
subset_metadata$is_MSM <- subset_metadata$MSM=="MSM"
ds.dat <- DESeqDataSetFromMatrix(countData=gns.only.subset, 
                                 colData=subset_metadata,
                                 design=~is_MSM)

# filter low count genes
keep <- rowSums(counts(ds.dat)) >= 10
ds.dat <- ds.dat[keep,]

dds <- DESeq(ds.dat)
res <- results(dds,  
               pAdjustMethod="fdr")
res.ape <- lfcShrink(dds=dds, coef=2, type="apeglm")

res <- res[order(res$padj),] 
res.ape <- res.ape[order(res.ape$padj),] 

res %>% as.data.frame() %>% head(n=10) %>% 
    knitr::kable(caption="Top DEGs")
res.ape %>% as.data.frame() %>% head(n=10) %>% 
    knitr::kable(caption="Top DEGs with LFC shrinkage applied")
```

### MA Plot {.tabset}

#### Apeglm shrunken LFC
```{r ma_shrunk.msm, fig.height=6}
plotMA(res.ape, ylim=c(-2,2))

```

#### Raw LFC
```{r ma_raw.msm, fig.height=6}
plotMA(res, ylim=c(-2,2))

```


### Volcano plot {.tabset}
Plots show unadjusted p values. The bonferroni-adjusted p value threshold is between 5 and 6 on the y axis, and if we do a Benjamini-Hochberg correction, again only genes with p < 10^-5 are significant. 

To the left indicates higher expression in MSM individuals, and to the right indicates higher expression in non-MSM individuals.

#### Apeglm shrunken LFC
```{r DESeq_volcano_shrunk.msm, fig.height=8}
EnhancedVolcano(res.ape,
                lab = rownames(res.ape),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```

#### Raw LFC
```{r DESeq_volcano_raw.msm, fig.height=8}
EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```


### pathfindR to map gene ids to pathways
PathfindR errors because it can't really do anything with the 2 differentially expressed genes.
```{r pathfindR.msm, eval=FALSE}
# Grab the relevant columns from the DESeq2 results
pfr.in <- res.ape[,c("log2FoldChange", "padj")] %>% as.data.frame()
pfr.in$Geneid <- rownames(pfr.in)

# order columns to be "Geneid", "log2FoldChange", "padj"
pfr.in <- pfr.in[,c(3,1,2)]

# remove NA p values
pfr.in <- pfr.in %>% filter(is.na(padj)==F)

pfr.res <- run_pathfindR(pfr.in, output_dir="pathfindR",
                         visualize_enriched_terms=F, plot_enrichment_chart=F)

```

### GSEA {.tabset}
```{r}
res2 <- res.ape %>% as.data.frame() %>%
  arrange(pvalue, log2FoldChange) %>%
  select(log2FoldChange) %>% 
  na.omit()

res2$SYMBOL <- rownames(res2)
# create a named vector of the gene stats
ranks <- tibble::deframe(res2[c("SYMBOL", "log2FoldChange")])

```
#### Hallmark
```{r gsea.msm, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)


fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()
```

#### Curated Gene Sets
```{r gsea.c2.msm, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.c2, stats=ranks)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)


```

#### Immune
```{r gsea.immune.msm, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.c7.immune, stats=ranks)


fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)
```

#### Cell types
```{r gsea.celltype.msm, fig.height=10}
fgseaRes <- fgsea(pathways=pathways.celltype, stats=ranks)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  filter(!is.na(log2err)) %>%
  arrange(padj) %>% 
  head(20) %>%
  knitr::kable()

hist(fgseaResTidy$padj)

```


# Deconvolution
fetch and gunzip the signature matrix from https://ftp.ncbi.nlm.nih.gov/geo/series/GSE116nnn/GSE116222/suppl/GSE116222%5FExpression%5Fmatrix.txt.gz
Infor about that matrix can be found here: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116222

```{r deconvolution}
abis <- immunedeconv::deconvolute(gns.only.tpm, method="abis")

#immunedeconv::deconvolute_cibersort(gns.only.tpm, arrays=F)

# es <- immunedeconv::deconvolute(gns.only.tpm, method="estimate")

```
