---
title: "Host Transcriptome"
author: "John Sterrett"
date: "2023-10-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load("tidyverse",
               "ggplot2",
               "DESeq2",
               "EnhancedVolcano",
               "pathfindR",
               "fgsea")

# set ggplot theme
theme_set(theme_bw())
```

# Load data
```{r}
gn <- data.table::fread("../seq.f0.0.r0.0.host/counts.txt", data.table=F)
colnames(gn) <- colnames(gn) %>% gsub(pattern="seq.f0.0.r0.0.host/",
                                      replacement="")
colnames(gn) <- colnames(gn) %>% gsub(pattern=".bam",
                                      replacement="")
rownames(gn) <- gn$Geneid
gn$Geneid <- NULL

gns.only <- gn[!colnames(gn) %in% c("Chr", "Start", "End", "Strand", "Length")] %>%
    as.matrix()



metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)
```

# Convert to transcripts per million (tpm)
```{r}
# https://support.bioconductor.org/p/91218/
# convert to TPM
tpm <- function(counts,len) {
  x <- counts/len
  return(t(t(x)*1e6/colSums(x)))
}

gns.only.tpm <- tpm(gns.only, gn$Length)
gns.only.tpm.pseudo <- tpm(gns.only+1, gn$Length)

write.csv(gns.only.tpm, "../seq.f0.0.r0.0.host//host_gene_counts_tpm.csv")

```

# Ordination {.tabset}
```{r}
tpm.aitch <- robCompositions::aDist(t(gns.only.tpm.pseudo))
tpm.aitch.pcoa <- ape::pcoa(tpm.aitch)


tpm.aitch.pcoa.dat <- merge(tpm.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")

```

## Plot by HIV Status {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

Does host transcriptome correlate with metatranscriptome and/or function?

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4)
```

## Plot by Cohort {.tabset}
### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4)
```

## Plot by storage method {.tabset}
```{r}
# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

tpm.aitch.pcoa.dat$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
tpm.aitch.pcoa.dat[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
tpm.aitch.pcoa.dat[is.na(tpm.aitch.pcoa.dat$storage_method), "storage_method"] <- "RNALater"

```

### Sample ID labeled
```{r warning=F}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
```

### Sample ID not labeled
```{r}
tpm.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4)
```

# Differential Expressions
## Prep data
```{r DESeq_prep}
metadata <- merge(metadata,
                  tpm.aitch.pcoa.dat[c("Row.names", "storage_method")], 
                  by.x="Sample_ID", by.y="Row.names")

get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

metadata$timepoint <- get_timepoint(metadata$Sample_ID)

subset_metadata <- metadata %>% filter(storage_method=="RNALater",
                                       timepoint==1)
gns.only.subset <- gns.only[,subset_metadata$Sample_ID]
```

## HIV Status
### Run DESeq2
```{r DESeq_run}
ds.dat <- DESeqDataSetFromMatrix(countData=gns.only.subset, 
                                 colData=subset_metadata,
                                 design=~HIV_Status)

# filter low count genes
keep <- rowSums(counts(ds.dat)) >= 10
ds.dat <- ds.dat[keep,]

dds <- DESeq(ds.dat)
res <- results(dds, contrast=c("HIV_Status", "Positive", "Negative"), 
               pAdjustMethod="fdr")
res.ape <- lfcShrink(dds=dds, coef=2, type="apeglm")

res <- res[order(res$padj),] 
res.ape <- res.ape[order(res.ape$padj),] 

res %>% as.data.frame() %>% head(n=5)
res.ape %>% as.data.frame() %>% head(n=5)
```

### Volcano plot {.tabset}
#### Apeglm shrunken LFC
```{r DESeq_volcano_shrunk}
EnhancedVolcano(res.ape,
                lab = rownames(res.ape),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```

#### Raw LFC
```{r DESeq_volcano_raw}
EnhancedVolcano(res,
                lab = rownames(res),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05)

```


### pathfindR to map gene ids to pathways
```{r pathfindR, eval=FALSE}
# Grab the relevant columns from the DESeq2 results
pfr.in <- res[,c("log2FoldChange", "padj")] %>% as.data.frame()
pfr.in$Geneid <- rownames(pfr.in)

# order columns to be "Geneid", "log2FoldChange", "padj"
pfr.in <- pfr.in[,c(3,1,2)]

# remove NA p values
pfr.in <- pfr.in %>% filter(is.na(padj)==F)

pfr.res <- run_pathfindR(pfr.in, output_dir="pathfindR",
                         visualize_enriched_terms=F, plot_enrichment_chart=F)

```

### GSEA
```{r gsea, fig.height=10}
res2 <- res %>% as.data.frame() %>%
  select(stat) %>% 
  na.omit() %>% 
  arrange(stat)

res2$SYMBOL <- rownames(res2)
# create a named vector of the gene stats
ranks <- tibble::deframe(res2[c("SYMBOL", "stat")])
# read in pathways, https://www.gsea-msigdb.org/gsea/msigdb
pathways.hallmark <- gmtPathways("../h.all.v2023.2.Hs.symbols.gmt")

fgseaRes <- fgsea(pathways=pathways.hallmark, stats=ranks)


fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  knitr::kable()

ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()
```
