---
title: "Microbe Functional"
author: "John Sterrett"
date: "2023-10-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load("tidyverse",
               "ggplot2",
               "HoMiStats")
# set ggplot theme
theme_set(theme_bw())
```

# Load Data
```{r}
gf <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies.tsv",
                        data.table=F)

rownames(gf) <- gf$`# Gene Family`
gf$`# Gene Family` <- NULL
gf <- gf %>% mutate_all(as.numeric)
colnames(gf) <- colnames(gf) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")

gf.tax.rows <- rownames(gf) %>% grepl(pattern="\\|")
gf.notax <- gf[!gf.tax.rows,]
gf.tax <- gf[gf.tax.rows,]

# GO terms
gf.go <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_go_named.tsv",
                        data.table=F)

rownames(gf.go) <- gf.go$`# Gene Family`
gf.go$`# Gene Family` <- NULL
gf.go <- gf.go %>% mutate_all(as.numeric)
colnames(gf.go) <- colnames(gf.go) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.go.tax.rows <- rownames(gf.go) %>% grepl(pattern="\\|")
gf.go.notax <- gf.go[!gf.go.tax.rows,]
gf.go.tax <- gf.go[gf.go.tax.rows,]


# Pfams
gf.pfam <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_pfam_named.tsv",
                        data.table=F)

rownames(gf.pfam) <- gf.pfam$`# Gene Family`
gf.pfam$`# Gene Family` <- NULL
gf.pfam <- gf.pfam %>% mutate_all(as.numeric)
colnames(gf.pfam) <- colnames(gf.pfam) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.pfam.tax.rows <- rownames(gf.pfam) %>% grepl(pattern="\\|")
gf.pfam.notax <- gf.pfam[!gf.pfam.tax.rows,]
gf.pfam.tax <- gf.pfam[gf.pfam.tax.rows,]


# KO
gf.ko <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_ko_named.tsv",
                        data.table=F)

rownames(gf.ko) <- gf.ko$`# Gene Family`
gf.ko$`# Gene Family` <- NULL
gf.ko <- gf.ko %>% mutate_all(as.numeric)
colnames(gf.ko) <- colnames(gf.ko) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.ko.tax.rows <- rownames(gf.ko) %>% grepl(pattern="\\|")
gf.ko.notax <- gf.ko[!gf.ko.tax.rows,]
gf.ko.tax <- gf.ko[gf.ko.tax.rows,]


# RXN EC
gf.rxn <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_rxn_named.tsv",
                        data.table=F)

rownames(gf.rxn) <- gf.rxn$`# Gene Family`
gf.rxn$`# Gene Family` <- NULL
gf.rxn <- gf.rxn %>% mutate_all(as.numeric)
colnames(gf.rxn) <- colnames(gf.rxn) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.rxn.tax.rows <- rownames(gf.rxn) %>% grepl(pattern="\\|")
gf.rxn.notax <- gf.rxn[!gf.rxn.tax.rows,]
gf.rxn.tax <- gf.rxn[gf.rxn.tax.rows,]

# Metadata
metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)

# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

metadata$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
metadata[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
metadata[is.na(metadata$storage_method), "storage_method"] <- "RNALater"

metadata[is.DM, "Participant"] <- paste0("DM_", number[is.DM])
metadata[!is.DM, "Participant"] <- paste0("LG_", number[!is.DM])

metadata %>% 
    group_by(Participant) %>% 
    summarise(MSM=dplyr::first(MSM)) %>%
    select(MSM) %>%
    as.data.frame() %>%
    mutate(MSM=as.factor(MSM)) %>%
    summary()
```

# Unmapped
```{r}
sample.sums <- colSums(gf)
percent.unmapped <- as.numeric((gf["UNMAPPED",]/sample.sums) * 100)
names(percent.unmapped) <- colnames(gf)

hist(percent.unmapped)

knitr::kable(percent.unmapped, col.names="Percent unmapped")
```

# PCoA {.tabset}

## Aitchison distance {.tabset}
```{r calc_dist}
notax.aitch <- robCompositions::aDist(t(gf.notax)+1)
tax.aitch <- robCompositions::aDist(t(gf.tax)+1)
```

### No-tax by HIV {.tabset}
#### Labeled
```{r pcoa_aitch_HIV}
notax.aitch.pcoa <- ape::pcoa(notax.aitch)

notax.aitch.pcoa.dat <- merge(notax.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

#### Not Labeled

```{r pcoa_aitch_HIV_nl}
notax.aitch.pcoa <- ape::pcoa(notax.aitch)

notax.aitch.pcoa.dat <- merge(notax.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4)
    
```

### No-tax by Cohort {.tabset}
#### Labeled
```{r pcoa_aitch_Cohort}
notax.aitch.pcoa.dat <- merge(notax.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

#### Not Labeled
```{r pcoa_aitch_Cohort_nl}
notax.aitch.pcoa.dat <- merge(notax.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4)
    
```

### No-tax by storage method {.tabset}
#### Labeled
```{r pcoa_aitch_storage}
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

#### Not Labeled
```{r pcoa_aitch_storage_nl}
notax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4)
    
```


### With tax by HIV
```{r pcoa_aitch_HIV_tax}
tax.aitch.pcoa <- ape::pcoa(tax.aitch)

tax.aitch.pcoa.dat <- merge(tax.aitch.pcoa$vectors, metadata,
                              by.x="row.names", by.y="Sample_ID")
tax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=HIV_Status, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

### With tax by Cohort
```{r pcoa_aitch_Cohort_tax}
tax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=MSM, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

### With tax by storage method
```{r pcoa_aitch_storage_tax}
tax.aitch.pcoa.dat %>%
    ggplot(mapping=aes(Axis.1, Axis.2, color=storage_method, label=Row.names)) +
    geom_point(size=4) +
    ggrepel::geom_label_repel(color="black")
    
```

## Distance-based statistics
### Prep
We'll need to only grab timepoint 1 data to avoid longitudinal sampling from the same individual
```{r}
get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

# subset the distance matrix to only time 1
index.t1 <- which(get_timepoint(names(notax.aitch))==1)
names.t1 <- names(notax.aitch)[index.t1]
t1.notax.aitch <- usedist::dist_subset(notax.aitch, names.t1)

# Subset the metadata
t1.notax.aitch.pcoa.dat <- notax.aitch.pcoa.dat[index.t1,]

# Make sure it worked
if(sum(t1.notax.aitch.pcoa.dat$Row.names==names(t1.notax.aitch)) != nrow(t1.notax.aitch.pcoa.dat)){
    stop("order of dist matrix and df are not the same")
}
```

### Run PERMANOVA on timepoint 1 only as a function of storage method
```{r}
vegan::adonis2(t1.notax.aitch ~ t1.notax.aitch.pcoa.dat$storage_method, 
               permutations=10^5)

```


# Differential Expression using HoMiStats {.tabset}
## Setup
```{r DEsetup}
metadata$HIV_Status_num <- metadata$HIV_Status %>%
    as.factor() %>%
    as.numeric() - 1

metadata$Participant_num <- metadata$Participant %>%
    as.factor() %>%
    as.numeric() - 1

metadata$MSM_bin <- (metadata$MSM=="MSM") %>%
    as.numeric()


get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

metadata$Timepoint <- get_timepoint(metadata$Sample_ID)
metadata.nonlong <- metadata %>%
    filter(Timepoint==1)

metadata.nonlong.nopbs <- metadata %>%
    filter(Timepoint==1,
           storage_method!="PBS")
```

## HIV {.tabset}
### Pfams
```{r homistats_pfam, warning=F}
gf.pfam.notax.rel <- t(gf.pfam.notax)/colSums(gf.pfam.notax)
gf.pfam.notax.rel.nonlong.nopbs <- gf.pfam.notax.rel[metadata.nonlong.nopbs$Sample_ID,]

pfam.DE <- run_mtxDE("HIV_Status_num", 
                   feature.table=gf.pfam.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

pfam.DE %>% filter(term=="var1",
                 q <= 0.2) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### KO
```{r homistats_ko, warning=F}
gf.ko.notax.rel <- t(gf.ko.notax)/colSums(gf.ko.notax)
gf.ko.notax.rel.nonlong.nopbs <- gf.ko.notax.rel[metadata.nonlong.nopbs$Sample_ID,]

ko.DE <- run_mtxDE("HIV_Status_num", 
                   feature.table=gf.ko.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

ko.DE %>% filter(term=="var1",
                 q <= 0.2) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### GO terms
```{r homistats_go, warning=F}
gf.go.notax.rel <- t(gf.go.notax)/colSums(gf.go.notax)
gf.go.notax.rel.nonlong.nopbs <- gf.go.notax.rel[metadata.nonlong.nopbs$Sample_ID,]

go.DE <- run_mtxDE("HIV_Status_num", 
                   feature.table=gf.go.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

go.DE %>% filter(term=="var1",
                 q <= 0.05) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### RXN numbers
```{r homistats_rxn, warning=F}
gf.rxn.notax.rel <- t(gf.rxn.notax)/colSums(gf.rxn.notax)
gf.rxn.notax.rel.nonlong.nopbs <- gf.rxn.notax.rel[metadata.nonlong.nopbs$Sample_ID,]

rxn.DE <- run_mtxDE("HIV_Status_num", 
                   feature.table=gf.rxn.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

rxn.DE %>% filter(term=="var1",
                 q <= 0.05) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


## MSM {.tabset}
### Pfams
```{r homistats_pfam_msm, warning=F}
pfam.DE.msm <- run_mtxDE(formula="MSM_bin", 
                   feature.table=gf.pfam.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

pfam.DE.msm %>% filter(term=="var1",
                 q <= 0.2) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### KO
```{r homistats_ko_msm, warning=F}
ko.DE.msm <- run_mtxDE("MSM_bin", 
                   feature.table=gf.ko.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

ko.DE.msm %>% filter(term=="var1",
                 q <= 0.2) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### GO terms
```{r homistats_go_msm, warning=F}
go.DE.msm <- run_mtxDE("MSM_bin", 
                   feature.table=gf.go.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

go.DE.msm %>% filter(term=="var1",
                 q <= 0.05) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


### RXN numbers
```{r homistats_rxn_msm, warning=F}
rxn.DE.msm <- run_mtxDE("MSM_bin", 
                   feature.table=gf.rxn.notax.rel.nonlong.nopbs, 
                   metadata=metadata.nonlong.nopbs,
                   sampleID="Sample_ID", 
                   reg.method="zibr",
                   ncores=8, 
                   show_progress=FALSE)

rxn.DE.msm %>% filter(term=="var1",
                 q <= 0.05) %>%
    arrange(q, descending=F) %>%
    select(c(feature,
             estimate, 
             q)) %>% 
    knitr::kable()

```


# Pathways
## Load and clean
```{r}
paths <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_pathabundance.tsv")

rownames(paths) <- paths$`# Pathway`
paths$`# Pathway` <- NULL
colnames(paths) <- colnames(paths) %>% gsub(pattern="_Abundance", 
                                            replacement="")

paths.tax.rows <- grepl("\\|", rownames(paths))

paths.tax <- paths[paths.tax.rows,]
rownames(paths.tax) <- rownames(paths)[paths.tax.rows]
paths.tax.rel <- t(t(paths.tax)/colSums(paths.tax))

paths.notax <- paths[!paths.tax.rows,]
rownames(paths.notax) <- rownames(paths)[!paths.tax.rows]
paths.notax.rel <- t(t(paths.notax)/colSums(paths.notax))
rownames(paths.notax.rel) <- rownames(paths.notax)
```

## Top Paths
```{r}
order.of.pathways <- order(rowSums(paths.notax.rel), decreasing=T)

path.notax.rel.sorted <- paths.notax.rel[order.of.pathways,] 
rownames(path.notax.rel.sorted) <- rownames(paths.notax.rel)[order.of.pathways]

pathway.means.sorted <- path.notax.rel.sorted %>%
    rowMeans() 

(pathway.means.sorted*100) %>% 
    round(digits=2) %>%
    head(n=20) %>%
    knitr::kable()

integrated.p.m.s <- (pathway.means.sorted[!(names(pathway.means.sorted) 
                       %in% c("UNMAPPED", "UNINTEGRATED"))]) 
((integrated.p.m.s/sum(integrated.p.m.s)) * 100) %>% 
    round(digits=2) %>%
    head(n=20) %>%
    knitr::kable()

```
