---
title: "Targeted Host-Microbe gene expression"
author: "John Sterrett"
date: "2024-04-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load("tidyverse",
               "ggplot2",
               "HoMiStats")
# set ggplot theme
theme_set(theme_bw())
```

# Load Data
```{r}
gf <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies.tsv",
                        data.table=F)

rownames(gf) <- gf$`# Gene Family`
gf$`# Gene Family` <- NULL
gf <- gf %>% mutate_all(as.numeric)
colnames(gf) <- colnames(gf) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")

gf.tax.rows <- rownames(gf) %>% grepl(pattern="\\|")
gf.notax <- gf[!gf.tax.rows,]
gf.tax <- gf[gf.tax.rows,]

# KO
gf.ko <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_ko_named.tsv",
                        data.table=F)

rownames(gf.ko) <- gf.ko$`# Gene Family`
gf.ko$`# Gene Family` <- NULL
gf.ko <- gf.ko %>% mutate_all(as.numeric)
colnames(gf.ko) <- colnames(gf.ko) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.ko.tax.rows <- rownames(gf.ko) %>% grepl(pattern="\\|")
gf.ko.notax <- gf.ko[!gf.ko.tax.rows,]
gf.ko.tax <- gf.ko[gf.ko.tax.rows,]


# RXN EC
gf.rxn <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_rxn_named.tsv",
                        data.table=F)

rownames(gf.rxn) <- gf.rxn$`# Gene Family`
gf.rxn$`# Gene Family` <- NULL
gf.rxn <- gf.rxn %>% mutate_all(as.numeric)
colnames(gf.rxn) <- colnames(gf.rxn) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")


gf.rxn.tax.rows <- rownames(gf.rxn) %>% grepl(pattern="\\|")
gf.rxn.notax <- gf.rxn[!gf.rxn.tax.rows,]
gf.rxn.tax <- gf.rxn[gf.rxn.tax.rows,]

# Gut metabolic modules
gmms <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/Gut_metabolic_modules.csv",
                        data.table=F)
rownames(gmms) <- gmms$V1
gmms$V1 <- NULL

# Host genes

gn <- data.table::fread("../seq.f0.0.r0.0.host/counts.txt", data.table=F)
colnames(gn) <- colnames(gn) %>% gsub(pattern="seq.f0.0.r0.0.host/",
                                      replacement="")
colnames(gn) <- colnames(gn) %>% gsub(pattern=".bam",
                                      replacement="")
rownames(gn) <- gn$Geneid
gn$Geneid <- NULL

gns.only <- gn[!colnames(gn) %in% c("Chr", "Start", "End", "Strand", "Length")] %>%
    as.matrix()

# Metadata
metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)

# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

metadata$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
metadata[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
metadata[is.na(metadata$storage_method), "storage_method"] <- "RNALater"

metadata[is.DM, "Participant"] <- paste0("DM_", number[is.DM])
metadata[!is.DM, "Participant"] <- paste0("LG_", number[!is.DM])

metadata %>% 
    group_by(Participant) %>% 
    summarise(MSM=dplyr::first(MSM)) %>%
    select(MSM) %>%
    as.data.frame() %>%
    mutate(MSM=as.factor(MSM)) %>%
    summary()

get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

metadata$timepoint <- get_timepoint(metadata$Sample_ID)
```

# Subset data to create fully non-longitudinal dataset
```{r}
subset.metadata <- metadata %>% filter(storage_method=="RNALater",
                                       timepoint==1)
gf.ko.tax.subset <- gf.ko.tax[,subset.metadata$Sample_ID]
gmms.subset <- gmms[,subset.metadata$Sample_ID]

gns.only.subset <- gns.only[,subset.metadata$Sample_ID]

```

# Unmapped
```{r}
sample.sums <- colSums(gf)
percent.unmapped <- as.numeric((gf["UNMAPPED",]/sample.sums) * 100)
names(percent.unmapped) <- colnames(gf)

hist(percent.unmapped)

knitr::kable(percent.unmapped, col.names="Percent unmapped")
```

# Butyrate

Individual genes involved directly in butyrate synthesis are expressed at very low TPM and are sparse.
```{r}
buk <- rownames(gf.ko.notax) %>% grep(pattern="K00929")
atoD <- rownames(gf.ko.notax) %>% grep(pattern="K01034")

gf.ko.notax[c(buk, atoD),] %>% t() %>% summary() %>% knitr::kable()
```

At the gut metabolic module level, transcripts for butyrate production are consistently detected across samples. However, these two pathways are identical (the overlapping KO is K00248, butyryl-CoA dehydrogenase, which takes us farther from butyrate...)
```{r}
gmms["butyrate production I",] %>% unlist() %>% hist()
gmms["butyrate production II",] %>% unlist() %>% hist()

```

## T cells {.tabset}
Differential expression analysis on the targeted GO terms. No real results

### Setup
```{r function_setup}

get_go_term_human_genes <- function(go.term){
    ensembl <- biomaRt::useMart("ensembl",dataset="hsapiens_gene_ensembl") #uses human ensembl annotations
    #gets gene symbol, transcript_id and go_id for all genes annotated with GO:0007507
    gene.data <- biomaRt::getBM(attributes=c('hgnc_symbol', 'ensembl_transcript_id', 'go_id'),
                                filters='go', values=go.term, mart=ensembl)
    
    targeted.genes <- unique(gene.data$hgnc_symbol)
    return(targeted.genes)
}

go_targeted_diffex <- function(go.term, host.genes, microbial.genes, microbial.gene, verbose=T){
    targeted.genes <- get_go_term_human_genes(go.term)

    # filter empty strings
    targeted.genes <- targeted.genes[nzchar(targeted.genes)]
   
    if(verbose){
        print(targeted.genes)
    }
    
    targeted.data <- host.genes[targeted.genes,]
    if(verbose){
        s1 <- paste0(length(targeted.genes), " genes identified from GO term")
        s2 <- paste0(nrow(targeted.data), " genes in DESeq2 dataset")
        print(s1)
        print(s2)
    }
    
    ds <- DESeq2::DESeqDataSetFromMatrix(
                                 countData=targeted.data, 
                                 colData=microbial.genes,
                                 design=as.formula(paste0("~",microbial.gene)))
    dds <- DESeq2::DESeq(ds)
    
    res <- DESeq2::results(dds, pAdjustMethod="fdr")
    return(res)
}
```

### 0042110 - T cell activation
```{r}
res <- go_targeted_diffex("GO:0042110", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="butyrate.production.I")


summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")
```


### 0050863 - Regulation of T cell activation
```{r}
res <- go_targeted_diffex("GO:0050863", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="butyrate.production.I")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")

plot(gns.only.subset["CD4",], gmms.subset["butyrate production I",])
```

### 0045066 - Regulatory T cell differentiation
```{r}
res <- go_targeted_diffex("GO:0045066", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="butyrate.production.I")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")
```

### 0046643 - Regulation of gamma delta T cell activation
Not run because 0 genes from this are in our dataset
```{r, eval=F}
res <- go_targeted_diffex("GO:0046643", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="butyrate.production.I")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")
```


# Mucin degradation

## Mucus secretion

### 0070701

```{r}
res <- go_targeted_diffex("GO:0070701", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="mucin.degradation")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")

#plot(gns.only.subset["CLDN18",], gmms.subset["mucin degradation",])

```

### All Muc genes

```{r}
muc.genes <- rownames(gns.only.subset)[grepl(pattern="Muc", x=rownames(gns.only.subset), 
                                ignore.case=T)]

```


## Tight junctions {.tabset}
A whole lotta nothing...

### 0070160-tight junctions
```{r}
res <- go_targeted_diffex("GO:0070160", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="mucin.degradation")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")

plot(gns.only.subset["CLDN18",], gmms.subset["mucin degradation",])

```

### 0120192-tight junction assembly
```{r}
res <- go_targeted_diffex("GO:0120192", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="mucin.degradation")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")
```

### 0120193-tight junction organization
```{r}
res <- go_targeted_diffex("GO:0120193", host.genes=gns.only.subset, 
                           microbial.genes=data.frame(t(gmms.subset)),
                           microbial.gene="mucin.degradation")

summary(res$pvalue)
res %>% as.data.frame() %>% 
    dplyr::filter(pvalue < 0.05) %>% 
    knitr::kable(caption="Significant genes")
```