---
title: "WGCNA"
author: "John Sterrett"
date: "2024-04-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# These are required for WGCNA but the standard CRAN install.packages doesn't 
# install them when installing WGCNA because they are bioc
bioc.deps <- c("impute", "GO.db", "preprocessCore")
for(package in bioc.deps){
    if(!require(package, character.only=T, quietly=T)){
        BiocManager::install(package)
    }    
}


if(!require("ggraph", character.only=T, quietly=T)){
    devtools::install_github("thomasp85/ggraph")
}

pacman::p_load("tidyverse",
               "ggplot2",
               "WGCNA",
               "HoMiStats")
```

# Read and clean data
```{r}
# microbe KOs
gf.ko <- data.table::fread("../seq.f0.0.r0.0.nonhost.humann/all_genefamilies_ko_named.tsv",
                        data.table=F)

rownames(gf.ko) <- gf.ko$`# Gene Family`
gf.ko$`# Gene Family` <- NULL
gf.ko <- gf.ko %>% mutate_all(as.numeric)
colnames(gf.ko) <- colnames(gf.ko) %>% 
    gsub(pattern="_Abundance-RPKs", replacement="")

gf.ko.tax.rows <- rownames(gf.ko) %>% grepl(pattern="\\|")
gf.ko.notax <- gf.ko[!gf.ko.tax.rows,]
gf.ko.tax <- gf.ko[gf.ko.tax.rows,]

# filter unmapped and ungrouped reads out
gf.ko.notax <- gf.ko.notax[!grepl(x=rownames(gf.ko.notax), 
                      pattern="UNMAPPED|UNGROUPED", perl=T),]
gf.ko.tax <- gf.ko.tax[!grepl(x=rownames(gf.ko.tax), 
                      pattern="UNMAPPED|UNGROUPED", perl=T),]

# Host genes
gn <- data.table::fread("../seq.f0.0.r0.0.host/counts.txt", data.table=F)
colnames(gn) <- colnames(gn) %>% gsub(pattern="seq.f0.0.r0.0.host/",
                                      replacement="")
colnames(gn) <- colnames(gn) %>% gsub(pattern=".bam",
                                      replacement="")
rownames(gn) <- gn$Geneid
gn$Geneid <- NULL

gns.only <- gn[!colnames(gn) %in% c("Chr", "Start", "End", "Strand", "Length")] %>%
    as.matrix()

# Metadata
metadata <- data.table::fread("../16s-data/16S_meta.txt",
                        data.table=F)

# Samples before DM 33 were stored in PBS, all others (including LG)
is.DM <- grepl("DM", metadata$Sample_ID)
number <- str_split(metadata$Sample_ID, pattern="_") %>% 
    lapply(function(x){x[2]}) %>% 
    unlist() %>% 
    as.numeric()

metadata$storage_method <- rep(NA, nrow(metadata)) %>% as.character()
metadata[((number < 33) & (is.DM)), "storage_method"] <- "PBS"
metadata[is.na(metadata$storage_method), "storage_method"] <- "RNALater"

metadata[is.DM, "Participant"] <- paste0("DM_", number[is.DM])
metadata[!is.DM, "Participant"] <- paste0("LG_", number[!is.DM])

metadata %>% 
    group_by(Participant) %>% 
    summarise(MSM=dplyr::first(MSM)) %>%
    select(MSM) %>%
    as.data.frame() %>%
    mutate(MSM=as.factor(MSM)) %>%
    summary()

get_timepoint <- function(sampleIDs) {
  split_strings <- str_split(sampleIDs, "_")
  # LG samples don't have a timepoint, so we'll just fill in "1" for them
  timepoints <- sapply(split_strings, function(x) {
    if (length(x) >= 3) {
      return(as.numeric(x[3]))
    } else {
      return(1)
    }
  })
  return(timepoints)
}

metadata$timepoint <- get_timepoint(metadata$Sample_ID)


```

# Subset data to create fully non-longitudinal dataset
```{r}
subset.metadata <- metadata %>% filter(storage_method=="RNALater",
                                       timepoint==1)
gf.ko.notax.subset <- gf.ko.notax[,subset.metadata$Sample_ID]

gns.only.subset <- gns.only[,subset.metadata$Sample_ID]

```

# Preprocess
## Host
```{r}
# Filter transcripts with fewer than 50 total reads across the dataset
gns.only.subset <- gns.only.subset[rowSums(gns.only.subset) > 50,]
# read into DESeq2 for normalization
ds.dat <- DESeq2::DESeqDataSetFromMatrix(gns.only.subset,
                               colData=subset.metadata,
                               design=~1)
dds <- DESeq2::DESeq(ds.dat)
# Normalize data
vsd <- DESeq2::getVarianceStabilizedData(dds)

# remove any genes below the 25th percentile of variance
rv_vsd <- matrixStats::rowVars(vsd)
summary(rv_vsd)
q25_vsd <- quantile(rv_vsd, .25)
gns.only.subset.normalized <- vsd[rv_vsd > q25_vsd, ]
print(paste("Host transcripts after filtering:", nrow(gns.only.subset.normalized)))
```

## Microbes

```{r}
# apply a TSS normalization (this is transcripts per million)
gf.ko.notax.subset.rel <- t(t(gf.ko.notax.subset)/colSums(gf.ko.notax.subset))
tol <- 1e-4
if(any(colSums(gf.ko.notax.subset.rel) < 1-tol | 
       colSums(gf.ko.notax.subset.rel) > 1+tol)){
    stop("Microbial KOs are not in relative abundances.")
}
# Filter transcripts seen less than 1 per 100000 transcripts
threshold.microbe.genes <- ncol(gf.ko.notax.subset.rel)*1e-6
gf.ko.notax.subset.rel.f <- gf.ko.notax.subset.rel[rowSums(gf.ko.notax.subset.rel) > threshold.microbe.genes,]

rv_gf.ko.notax.subset.rel.f <- matrixStats::rowVars(gf.ko.notax.subset.rel.f)
summary(rv_gf.ko.notax.subset.rel.f)
q25_rv_gf.ko.notax.subset.rel.f <- quantile(rv_gf.ko.notax.subset.rel.f, .25)
gf.ko.notax.subset.rel.f <- gf.ko.notax.subset.rel.f[rv_gf.ko.notax.subset.rel.f > q25_rv_gf.ko.notax.subset.rel.f, ]
print(paste("Microbial KOs after filtering:", nrow(gf.ko.notax.subset.rel.f)))
```


# WGCNA setup
```{r}
n.threads <- 16
allowWGCNAThreads(n.threads)

# ~16 GB RAM, per Langfelder's blog
# https://peterlangfelder.com/2018/11/25/blockwise-network-analysis-of-large-data/
max.block.size <- 20000
```

# WGCNA on host
```{r wgcna_host}
# Choose a set of soft-thresholding powers
powers <- c(c(1:10), seq(from = 12, to = 20, by = 2))

# Call the network topology analysis function
sft <- pickSoftThreshold(
  t(gns.only.subset.normalized),
  blockSize = max.block.size,
  powerVector = powers,
  verbose = 5
)

sft_df <- data.frame(sft$fitIndices) %>%
  dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)

ggplot(sft_df, aes(x=Power, y=model_fit, label=Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  # Plot the WGCNA recommended R^2 cutoff
  geom_hline(yintercept = 0.80, col = "red") +
  ylim(c(min(sft_df$model_fit), 1.05)) +
  xlab("Soft Threshold (power)") +
  ylab("Scale Free Topology Model Fit, signed R^2") +
  theme_bw()

chosen.power <- 20

bwnet <- blockwiseModules(t(gns.only.subset.normalized),
  maxBlockSize=max.block.size,
  TOMType="signed",
  power=chosen.power, # soft threshold for network construction
  numericLabels=TRUE, # Use numbers instead of colors for module labels
  randomSeed=1234
)
readr::write_rds(bwnet,
  file = file.path("host_wgcna_results.RDS")
)



mergedColors <- labels2colors(bwnet$colors)

plotDendroAndColors(
  bwnet$dendrograms[[1]],
  mergedColors[bwnet$blockGenes[[1]]],
  "Module colors",
  dendroLabels=FALSE,
  hang=0.03,
  addGuide=TRUE,
  guideHang=0.05 )

plotEigengeneNetworks(bwnet$MEs, "Eigengene Adjacency Heatmap", marHeatmap = c(3,4,2,2))
```

# WGCNA on microbes
```{r wgcna_microbes}
extra.powers <- c(powers, seq(22, 50, by=2))

# Call the network topology analysis function
sft <- pickSoftThreshold(
  t(gf.ko.notax.subset.rel.f),
  blockSize=max.block.size,
  powerVector=extra.powers,
  verbose=5
)

sft_df <- data.frame(sft$fitIndices) %>%
  dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)

ggplot(sft_df, aes(x=Power, y=model_fit, label=Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  # Plot the WGCNA recommended R^2 cutoff
  geom_hline(yintercept = 0.80, col = "red") +
  ylim(c(min(sft_df$model_fit), 1.05)) +
  xlab("Soft Threshold (power)") +
  ylab("Scale Free Topology Model Fit, signed R^2") +
  theme_bw()

chosen.power <- 30

bwnet.microbes <- blockwiseModules(t(gf.ko.notax.subset.rel.f),
  maxBlockSize=max.block.size,
  TOMType="signed",
  power=chosen.power, # soft threshold for network construction
  numericLabels=TRUE, # Use numbers instead of colors for module labels
  randomSeed=1234
)

readr::write_rds(bwnet.microbes,
  file = file.path("microbes_wgcna_results.RDS")
)


mergedColors <- labels2colors(bwnet.microbes$colors)

plotDendroAndColors(
  bwnet.microbes$dendrograms[[1]],
  mergedColors[bwnet.microbes$blockGenes[[1]]],
  "Module colors",
  dendroLabels=FALSE,
  hang=0.03,
  addGuide=TRUE,
  guideHang=0.05)

plotEigengeneNetworks(bwnet.microbes$MEs, "Eigengene Adjacency Heatmap", marHeatmap = c(3,4,2,2))
```

# WGCNA on all
```{r wgcna_all}
extra.powers <- c(powers, seq(22, 50, by=2))

both.datasets <- merge(t(gf.ko.notax.subset.rel.f), t(gns.only.subset.normalized), by="row.names")
rownames(both.datasets) <- both.datasets$Row.names
both.datasets$Row.names <- NULL
# Call the network topology analysis function

sft <- pickSoftThreshold(
  both.datasets,
  blockSize=max.block.size,
  powerVector=extra.powers,
  verbose=5
)

sft_df <- data.frame(sft$fitIndices) %>%
  dplyr::mutate(model_fit = -sign(slope) * SFT.R.sq)

ggplot(sft_df, aes(x=Power, y=model_fit, label=Power)) +
  geom_point() +
  geom_text(nudge_y = 0.05) +
  # Plot the WGCNA recommended R^2 cutoff
  geom_hline(yintercept = 0.80, col = "red") +
  ylim(c(min(sft_df$model_fit), 1.05)) +
  xlab("Soft Threshold (power)") +
  ylab("Scale Free Topology Model Fit, signed R^2") +
  theme_bw()

chosen.power <- 24

bwnet.all <- blockwiseModules(both.datasets,
  maxBlockSize=max.block.size,
  TOMType="signed",
  power=chosen.power, # soft threshold for network construction
  numericLabels=TRUE, # Use numbers instead of colors for module labels
  randomSeed=1234
)

readr::write_rds(bwnet.all,
  file = file.path("all_wgcna_results.RDS")
)


mergedColors <- labels2colors(bwnet.all$colors)

plotDendroAndColors(
  bwnet.all$dendrograms[[1]],
  mergedColors[bwnet.all$blockGenes[[1]]],
  "Module colors",
  dendroLabels=FALSE,
  hang=0.03,
  addGuide=TRUE,
  guideHang=0.05)

plotEigengeneNetworks(bwnet.all$MEs, "Eigengene Adjacency Heatmap", marHeatmap = c(3,4,2,2))


```


# SCNIC on microbes

Prep

```{r}
gf.ko.notax.subset.rel.f.biom <- biomformat::make_biom(gf.ko.notax.subset.rel.f)
if(!file.exists("gf.ko.notax.subset.rel.f.biom")){
    biomformat::write_biom(gf.ko.notax.subset.rel.f.biom, 
                       "gf.ko.notax.subset.rel.f.biom")
}
```


Setting min_r for the SCNIC_modules - need to pick a low value because the r 
values are quite low (max is .21)
```{bash, eval=F}
conda activate SCNIC
SCNIC_analysis.py within -i gf.ko.notax.subset.rel.f.biom -o SCNIC_within_output/ -m sparcc
SCNIC_analysis.py modules -i SCNIC_within_output/correls.txt -o SCNIC_modules_output/ --min_r .1 --table gf.ko.notax.subset.rel.f.biom
```

```{r}
scnic.biom <- biomformat::read_biom("SCNIC_modules_output/collapsed.biom") 
scnic.table <- biomformat::biom_data(scnic.biom) %>% 
    as.matrix() %>% 
    as.data.frame() %>%
    t()

data.table::fread("SCNIC_within_output/correls.txt") %>% 
    select("r") %>% 
    hist()


```


# Associations between host modules and microbes modules
## WGCNA-WGCNA
```{r}
bwnet.microbes.df <- bwnet.microbes$MEs
colnames(bwnet.microbes.df) <- paste0("microbes_", colnames(bwnet.microbes.df))

bwnet.df <- bwnet$MEs
colnames(bwnet.df) <- paste0("host_", colnames(bwnet.df))

homicorr.res.w.w <- HoMiStats::run_HoMiCorr(mtx=bwnet.microbes.df, host=bwnet$MEs, 
                                            reg.method="lm")

homicorr.res.w.w %>%
    filter(q < 0.05,
           (grepl(pattern="microbes", x=homicorr.res.w.w$feature) + 
            grepl(pattern="microbes", x=homicorr.res.w.w$term))==1) %>%
    arrange(estimate) %>%
    knitr::kable()

bwnet.modules <- data.frame(name=names(bwnet$colors), 
                            module=unname(bwnet$colors))
bwnet.modules <- bwnet.modules[order(bwnet.modules$module),]
write.csv(bwnet.modules, "host_wgcna_module_contents.csv")

bwnet.microbes.modules <- data.frame(name=names(bwnet.microbes$colors), 
                                     module=unname(bwnet.microbes$colors))
bwnet.microbes.modules <- bwnet.microbes.modules[order(bwnet.microbes.modules$module),]
write.csv(bwnet.microbes.modules, "microbe_wgcna_module_contents.csv") 

```

Network visualization of WGCNA-WGCNA modules
```{r}
old.par <- par(bg=NA)
graph.data.w.w <- homicorr.res.w.w %>%
    filter(q < 0.05) %>%
    dplyr::select(c(term, feature, statistic))

node.names <- unique(c(graph.data.w.w$term, graph.data.w.w$feature))
node.origin <- grepl("microbes", x=node.names)
node.data <- data.frame(node=node.names, microbial=node.origin)
node.data[node.origin, "transcriptome"] <- "Microbial"
node.data[!node.origin, "transcriptome"] <- "Host"
node.data[node.origin, "color"] <- "red3"
node.data[!node.origin, "color"] <- "blue3"

net <- igraph::graph_from_data_frame(graph.data.w.w, vertices=node.data, 
                                     directed=F)
igraph::E(net)$weight <- graph.data.w.w$statistic


ncolors <- length(unique(graph.data.w.w$statistic))
brbg <- rev(RColorBrewer::brewer.pal(ncolors, "BrBG"))
brbg_hcl <- c(colorRampPalette(c(brbg[1], brbg[6]))(51), 
    colorRampPalette(c(brbg[6], brbg[11]))(51)[-1])

pal <- scales::gradient_n_pal(colours = c("green","black","blue"),
      values= c(-42, 0, 42),)
# use data in a to generate palette-generated values in the data
igraph::E(net)$color <- pal(igraph::E(net)$weight)

c_scale <- viridis::viridis(length(unique(graph.data.w.w$statistic)))
igraph::E(net)$color <-brbg_hcl[order(graph.data.w.w$statistic)]


net <- igraph::simplify(net, remove.multiple=F, remove.loops=T) 
igraph::V(net)$colors <- node.data$color
igraph::V(net)$label <- NA


circle.layout <- igraph::layout_in_circle(net)
plot(net, layout=circle.layout) 
legend(x=-1, y=-0.85, c("Host", "Microbial"), pch=21,
       col="#777777", pt.bg=c("blue3", "red3"), pt.cex=2, cex=.8, bty="n", ncol=1)

dev.copy(png,'circle_w_w.png')
dev.off()
par(old.par)



non.signed.net <- net
igraph::E(non.signed.net)$weight <- abs(igraph::E(non.signed.net)$weight)

fr.layout <- igraph::layout_with_fr(non.signed.net)
plot(non.signed.net, layout=fr.layout)
legend(x="bottomleft", c("Host", "Microbial"), pch=21,
       col="#777777", pt.bg=c("blue3", "red3"), pt.cex=2, cex=.8, bty="n", ncol=1)


igraph::E(net)$Weight <- igraph::E(net)$weight
igraph::V(net)$Transcriptome <- igraph::V(net)$transcriptome

names(igraph::V(net))
s <- sort(names(igraph::V(net)))
net.sorted <- igraph::permute(net, match(igraph::V(net)$name, s))


ggraph(net.sorted, layout="circle") +
    geom_edge_link(aes(edge_color=Weight), linewidth=1, alpha=1) +   # add edges to the plot
    geom_node_point(aes(color=Transcriptome), size=12) +
    scale_edge_color_gradient2(low="forestgreen", high="brown",mid="grey",) +
    scale_color_manual(labels=c("Host", "Microbial"),
                       values=c("blue3", "red3")) +
    theme_void() +
    theme(legend.title=element_text(size=26), 
    legend.text=element_text(size=20))

ggsave("circle_network.png", bg="transparent", width=12, height=10)
```

## WGCNA-SCNIC
```{r, eval=FALSE}

scnic.table.clean <- janitor::clean_names(scnic.table)
rownames(scnic.table.clean) <- rownames(scnic.table)

homicorr.res.s.w <- HoMiStats::run_HoMiCorr(mtx=scnic.table.clean, host=bwnet.df, 
                                            reg.method="lm",
                                            zero_prop_from_formula=TRUE,
                                            ncores=32)

homicorr.res.s.w %>%
    filter(q < 0.05,
           (grepl(pattern="host", x=homicorr.res.s.w$feature) + 
            grepl(pattern="host", x=homicorr.res.s.w$term))==1
           ) %>%
    arrange(estimate) %>%
    knitr::kable()
```


# WGCNA on all
```{r}
bwnet.modules.all <- data.frame(name=names(bwnet.all$colors), 
                            module=unname(bwnet.all$colors))
bwnet.modules.all <- bwnet.modules.all[order(bwnet.modules.all$module),]
write.csv(bwnet.modules.all, "all_wgcna_module_contents.csv")
```


# Intra-module hubs
```{r}


```